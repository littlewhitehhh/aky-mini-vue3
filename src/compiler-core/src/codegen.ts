// import { isString } from "../../shared";
// import { NodeTypes } from "./ast";
// import { Create_ELEMENT_Vnode, helperMapName, To_Display_String } from "./runtimeHelpers";

// export function generate(ast) {
//   // let code = "";
//   const context: any = createCodegenContext();
//   const { push } = context;

//   // const VueBinging = "Vue";
//   // // const helpers = ["toDisplayString"];
//   // const aliasHelper = (s) => `${s}:_${s}`;
//   // push(`const ${ast.helpers.map(aliasHelper).join(",")}= ${VueBinging} `);
//   // push("\n");
//   // // code += "return";
//   // push("return ");

//   //重构
//   genFunctionPreamble(context, ast);

//   const functionName = "render";
//   const args = ["_cxt", "_cache"];
//   const signature = args.join(",");

//   push(`function ${functionName}(${signature}){`);
//   // code += ` function ${functionName}(${signature}){`;

//   console.log(ast);

//   // code += `return`;
//   push(`return `);
//   genNode(ast.codegenNode, context);
//   // code += "}";
//   push("}");

//   return {
//     // code: `return function render(_cxt,_cache,$props,$setup,$data,,$options){
//     //   return "hi"
//     // }`,
//     code: context.code,
//   };
// }
// function genNode(node: any, context: any) {
//   switch (node.type) {
//     case NodeTypes.TEXT:
//       genText(context, node);
//       break;
//     case NodeTypes.INTERPOLATION:
//       genInterpolation(node, context);
//       break;
//     case NodeTypes.SIMPLE_EXPRESSION:
//       genExpression(node, context);
//       break;
//     case NodeTypes.ELEMENT:
//       genElement(node, context);
//       break;
//     case NodeTypes.COMPOUND_EXPRESSION:
//       genCompoundExpression(node, context);
//       break;
//     default:
//       break;
//   }
// }
// function createCodegenContext() {
//   const context = {
//     code: "",
//     push(source) {
//       context.code += source;
//     },
//     helper(key) {
//       return `_${helperMapName[key]}`;
//     },
//   };
//   return context;
// }

// function genFunctionPreamble(context: any, ast: any) {
//   const VueBinging = "Vue";
//   // const helpers = ["toDisplayString"];
//   const { push } = context;
//   const aliasHelper = (s) => `${helperMapName[s]}:_${helperMapName[s]}`;

//   if (ast.helpers.length > 0) {
//     push(`const {${ast.helpers.map(aliasHelper).join(",")}}= ${VueBinging} `);
//   }

//   push("\n");
//   // // code += "return";
//   push("return ");
// }
// function genText(context, node) {
//   const { push } = context;
//   // code += ` "${node.content}"`;
//   push(`'${node.content}'`);
//   // return code;
// }

// function genInterpolation(node: any, context: any) {
//   const { push, helper } = context;
//   console.log(node);

//   push(`_${helper(To_Display_String)}(`);
//   genNode(node.content, context);
//   push(")");
// }

// function genExpression(node: any, context: any): any {
//   const { push } = context;

//   push(`${node.content}`);
// }

// function genElement(node: any, context: any) {
//   const { push, helper } = context;

//   const { tag, children, props } = node;

//   push(`${helper(Create_ELEMENT_Vnode)}(`);
//   genNodeList(genNullable([tag, props, children]), context);
//   // genNode(children, context);
//   // for (let i = 0; i < children.length; i++) {
//   //   const child = children[i];

//   //   genNode(child, context);
//   // }

//   push(")");
// }
// function genCompoundExpression(node: any, context: any) {
//   const { children } = node;
//   const { push } = context;
//   for (let i = 0; i < children.length; i++) {
//     const child = children[i];
//     if (isString(child)) {
//       push(child);
//     } else {
//       genNode(child, context);
//     }
//   }
// }
// function genNullable(args: any[]) {
//   return args.map((arg) => {
//     arg || "null";
//   });
// }

// function genNodeList(nodes, context) {
//   const { push } = context;
//   for (let i = 0; i < nodes.length; i++) {
//     const node = nodes[i];
//     if (isString(node)) {
//       push(node);
//     } else {
//       genNode(node, context);
//     }
//     if (i < nodes.length - 1) {
//       push(",");
//     }
//   }
// }

import { isString } from "../../shared";
import { NodeTypes } from "./ast";
import { Create_ELEMENT_Vnode, helperMapName, To_Display_String } from "./runtimeHelpers";

export function generate(ast) {
  const context = createCodegenContext();
  const { push } = context;

  genFunctionPreamble(ast, context);

  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(", ");

  push(`function ${functionName}(${signature}){`);
  push("return ");
  genNode(ast.codegenNode, context);
  push("}");

  return {
    code: context.code,
  };
}

function genFunctionPreamble(ast, context) {
  const { push } = context;
  const VueBinging = "Vue";
  const aliasHelper = (s) => `${helperMapName[s]}:_${helperMapName[s]}`;
  if (ast.helpers.length > 0) {
    push(`const { ${ast.helpers.map(aliasHelper).join(", ")} } = ${VueBinging}`);
  }
  push("\n");
  push("return ");
}

function createCodegenContext(): any {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
    helper(key) {
      return `_${helperMapName[key]}`;
    },
  };

  return context;
}

function genNode(node: any, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);
      break;

    default:
      break;
  }
}

function genCompoundExpression(node: any, context: any) {
  const { push } = context;
  const children = node.children;
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (isString(child)) {
      push(child);
    } else {
      genNode(child, context);
    }
  }
}

function genElement(node: any, context: any) {
  const { push, helper } = context;
  const { tag, children, props } = node;
  push(`${helper(Create_ELEMENT_Vnode)}(`);
  genNodeList(genNullable([tag, props, children]), context);
  push(")");
}

function genNodeList(nodes, context) {
  const { push } = context;

  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (isString(node)) {
      push(node);
    } else {
      genNode(node, context);
    }

    if (i < nodes.length - 1) {
      push(", ");
    }
  }
}

function genNullable(args: any) {
  return args.map((arg) => arg || "null");
}

function genExpression(node: any, context: any) {
  const { push } = context;
  push(`${node.content}`);
}

function genInterpolation(node: any, context: any) {
  const { push, helper } = context;
  push(`${helper(To_Display_String)}(`);
  genNode(node.content, context);
  push(")");
}

function genText(node: any, context: any) {
  const { push } = context;
  push(`'${node.content}'`);
}
