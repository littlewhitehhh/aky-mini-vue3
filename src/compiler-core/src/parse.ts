import { NodeTypes } from "./ast";

const enum TagType {
  Start,
  End,
}

export function baseParse(content: string) {
  //message
  const context = createParseContext(content); //{source:message}
  //重构
  return createRoot(parseChildren(context, [])); //
  // return {
  //   children: [
  //     {
  //       type: "interpolation",
  //       content: { type: "simple_expression", content: "message" },
  //     },
  //   ],
  // };

  // return createRoot([
  //   {
  //     type: "interpolation",
  //     content: { type: "simple_expression", content: "message" },
  //   },
  // ]);
}

function createParseContext(content): any {
  return {
    source: content,
  };
}
function createRoot(children) {
  return {
    children,
  };
}

function parseChildren(context, ancestors) {
  // ancestors 祖先
  const nodes: any = [];

  while (!isEnd(context, ancestors)) {
    let node;
    //解析插值{{}}
    if (context.source.startsWith("{{")) {
      node = parseInterpolation(context);
    } else if (context.source[0] === "<") {
      //element类型
      if (/[a-z]/.test(context.source[1])) {
        console.log("parse.element");
        node = parseElement(context, ancestors);
      }
    }
    // } else {
    //   //text类型
    //   node = parseText(context);
    // }
    if (!node) {
      node = parseText(context);
    }
    nodes.push(node);
  }
  return nodes;
  // return [
  //   {
  //     type: "interpolation",
  //     content: { type: "simple_expression", content: "message" },
  //   },
  // ];
}

function isEnd(context, ancestors) {
  //<div></div>

  //结束的条件
  //1、context.source 没有值的时候
  //2、当遇到结束标签时候

  //2

  const s = context.source;
  //</div>
  if (s.startsWith("</")) {
    for (let i = ancestors.length - 1; i >= 0; i--) {
      const tag = ancestors[i].tag;
      if (startsWithEndTagOpen(s, tag)) {
        return true;
      }
    }
  }

  // if (parentTag && s.startsWith(`</${parentTag}>`)) {
  //   return true;
  // }
  //1
  return !context.source;
}

//element类型parse
function parseElement(context: any, ancestors) {
  //implement    <div></div>
  // 1、解析tag
  // 2、删除处理完成的代码

  //1
  // const match: any = /^<([a-z]*)/i.exec(context.source); //
  // console.log(match); // [ '<div', 'div', index: 0, input: '<div></div>', groups: undefined ]
  // const tag = match[1];
  // console.log(tag); //div

  // //2  删除处理完成的代码
  // advanceBy(context, match[0].length);
  // console.log(context.source); //></div>
  // advanceBy(context, 1);
  // console.log(context.source); // </div>

  // return {
  //   type: NodeTypes.ELEMENT,
  //   tag,
  // };

  // 重构
  const element: any = parseTag(context, TagType.Start);
  ancestors.push(element);
  element.children = parseChildren(context, ancestors);
  ancestors.pop();

  // if (context.source.slice(2, 2 + element.tag.length) === element.tag) {

  //重构
  if (startsWithEndTagOpen(context.source, element.tag)) {
    parseTag(context, TagType.End);
  } else {
    throw new Error(`缺少结束标签:${element.tag}`);
  }

  console.log("-------", context.source);

  return element;
}

function parseTag(context: any, TagType) {
  //implement    <div></div>
  // 1、解析tag
  // 2、删除处理完成的代码

  //1
  const match: any = /^<\/?([a-z]*)/i.exec(context.source);
  console.log(match); // [ '<div', 'div', index: 0, input: '<div></div>', groups: undefined ]
  const tag = match[1];
  console.log(tag); //div

  // 2
  advanceBy(context, match[0].length); //></div>

  advanceBy(context, 1); //</div>

  if (tag === TagType.End) return;

  return {
    type: NodeTypes.ELEMENT,
    tag,
  };
}

//{{}}插值类型parse
function parseInterpolation(context) {
  console.log(context);

  const openDelimiter = "{{";
  const closeDelimiter = "}}";

  //{{message}}解析
  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length);

  advanceBy(context, openDelimiter.length);

  // context.source = context.source.slice(openDelimiter.length); // message}}
  console.log(context.source);

  const rawContentLength = closeIndex - openDelimiter.length;
  // const rawContent = context.source.slice(0, rawContentLength); // message
  const rawContent = parseTextData(context, rawContentLength);

  //边缘处理  利于 {{ message}}  双括号中有空格
  const content = rawContent.trim();
  console.log(content);

  // 例如{{message}}</div>   {{message}}处理完毕的 删除掉 剩下</div>
  // context.source = context.source.slice(closeDelimiter.length);
  advanceBy(context, closeDelimiter.length);
  console.log("context.source", context.source);

  return {
    type: NodeTypes.INTERPOLATION, //"interpolation"
    content: { type: NodeTypes.SIMPLE_EXPRESSION, content: content }, //"simple_expression"
  };
}

//TEXT类型parse
function parseText(context) {
  //1、获取content
  //2、推进代码  ： 删除已经处理的代码

  //1
  // const content = context.source.slice(0, context.source.length);
  // console.log(content);

  //2
  // advanceBy(context, content.length);
  // console.log(context.source);

  let endIndex = context.source.length;
  let endTokens = ["<", "{{"];

  for (let i = 0; i < endTokens.length; i++) {
    const index = context.source.indexOf(endTokens[i]);
    if (index !== -1 && endIndex > index) {
      //！==-1 证明没有插值
      endIndex = index;
    }
  }

  //重构
  const content = parseTextData(context, endIndex);
  console.log("content _______", content);

  return {
    type: NodeTypes.TEXT,
    content,
  };
}

//推进工具函数
function advanceBy(context: any, length: number) {
  context.source = context.source.slice(length);
}

//裁剪工具函数
function parseTextData(context, length) {
  const content = context.source.slice(0, length);

  advanceBy(context, length);
  return content;
}

function startsWithEndTagOpen(source, tag) {
  return source.startsWith("</") && source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase();
}
