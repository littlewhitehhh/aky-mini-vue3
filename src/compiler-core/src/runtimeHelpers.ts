export const To_Display_String = Symbol("toDisplayString");
export const Create_ELEMENT_Vnode = Symbol("createElementVnode");

export const helperMapName = {
  [To_Display_String]: "toDisplayString ",
  [Create_ELEMENT_Vnode]: "createElementVnode ",
};
