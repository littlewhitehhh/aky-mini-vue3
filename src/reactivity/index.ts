export function add(a, b) {
  return a + b;
}

export { ref, proxyRefs } from "./ref";
