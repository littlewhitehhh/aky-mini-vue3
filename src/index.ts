// mini-vue 的出口

export * from "./runtime-dom/index";

export * from "./reactivity/index";

import { baseCompile } from "./compiler-core/src";
import * as runtimeDom from "./runtime-dom";
import { registerRuntimeCompiler } from "./runtime-dom";

// render

function compoileToFunction(template) {
  const { code } = baseCompile(template);
  const render = new Function("Vue", code)(runtimeDom);
  return render;
}
registerRuntimeCompiler(compoileToFunction);
