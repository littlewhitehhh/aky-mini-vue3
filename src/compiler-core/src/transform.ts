import { NodeTypes } from "./ast";
import { To_Display_String } from "./runtimeHelpers";

export function transform(root, options = {}) {
  const context = createTransformContext(root, options);

  //1、遍历 ——深度优先搜索  递归
  // 2、修改text content
  traverseNode(root, context);

  createRootCodegen(root);

  root.helpers = [...context.helpers.keys()];
}

function createTransformContext(root: any, options: any) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1);
    },
  };

  return context;
}

function traverseNode(node: any, context) {
  console.log(node);
  //1.element

  const nodeTransforms = context.nodeTransforms;
  const onExitFns: any = [];
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    const onExit = transform(node, context);
    if (onExit) {
      onExitFns.push(onExit);
    }
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(To_Display_String);
      break;
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node, context);
    default:
      break;
  }
  let i = onExitFns.length;
  while (i--) {
    onExitFns[i]();
  }
  // traverseChildren(node, context);
}

function traverseChildren(node, context) {
  const children = node.children;

  for (let i = 0; i < children.length; i++) {
    const node = children[i];
    traverseNode(node, context);
  }
}
function createRootCodegen(root: any) {
  const child = root.children[0];

  if (child.type === NodeTypes.ELEMENT) {
    root.codegenNode = child.codegenNode;
  } else {
    root.codegenNode = root.children[0];
  }
}
