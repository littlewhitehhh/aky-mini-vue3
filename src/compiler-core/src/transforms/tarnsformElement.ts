import { createVNodeCall, NodeTypes } from "../ast";
import { Create_ELEMENT_Vnode } from "../runtimeHelpers";

export function transformElement(node, context) {
  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      context.helper(Create_ELEMENT_Vnode);
      // 中间处理层
      const vnodeTag = `"${node.tag}"`;
      //props
      let vnodeProps;

      // children
      const children = node.children;
      let vnodeChildren = children[0];

      // const vnodeElement = {
      //   type: NodeTypes.ELEMENT,
      //   tag: vnodeTag,
      //   prpos: vnodeProps,
      //   children: vnodeChildren,
      // };
      node.codegenNode = createVNodeCall(context, vnodeTag, vnodeProps, vnodeChildren);
    };
  }
}
