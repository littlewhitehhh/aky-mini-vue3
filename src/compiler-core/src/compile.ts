import { generate } from "./codegen";
import { baseParse } from "./parse";
import { transform } from "./transform";
import { transformElement } from "./transforms/tarnsformElement";
import { transformExpression } from "./transforms/transformExpression";
import { transformText } from "./transforms/transfromText";

export function baseCompile(template) {
  const ast: any = baseParse(template);
  transform(ast, {
    nodeTransforms: [transformExpression, transformElement, transformText],
  });
  console.log("ast_______", ast);

  return generate(ast);
}
