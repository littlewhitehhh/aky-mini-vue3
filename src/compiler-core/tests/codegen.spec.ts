import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformElement } from "../src/transforms/tarnsformElement";
import { transformExpression } from "../src/transforms/transformExpression";
import { transformText } from "../src/transforms/transfromText";

describe("codegen", () => {
  it("string", () => {
    const ast = baseParse("hi");
    transform(ast);
    const { code } = generate(ast);

    //快照 作用
    //1、抓bug
    // 2、有意的改变
    expect(code).toMatchSnapshot();
  });

  it("interpolation", () => {
    const ast = baseParse("{{message}}");

    transform(ast, {
      nodeTransforms: [transformExpression],
    });

    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("element", () => {
    const ast: any = baseParse("<div>hi,{{message}}</div>");

    transform(ast, {
      nodeTransforms: [transformExpression, transformElement, transformText],
    });
    console.log("ast_______", ast);

    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
});
