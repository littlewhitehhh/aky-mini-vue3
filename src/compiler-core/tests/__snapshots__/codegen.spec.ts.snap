// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"const { toDisplayString :_toDisplayString , createElementVnode :_createElementVnode  } = Vue
return function render(_ctx, _cache){return _createElementVnode (\\"div\\", null, 'hi,' + _toDisplayString (_ctx.message))}"
`;

exports[`codegen interpolation 1`] = `
"const { toDisplayString :_toDisplayString  } = Vue
return function render(_ctx, _cache){return _toDisplayString (_ctx.message)}"
`;

exports[`codegen string 1`] = `
"
return function render(_ctx, _cache){return 'hi'}"
`;
